{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "soi = ['kluyveromyces_marxianus', 'komagataella_phaffii', 'yarrowia_lipolytica']\n",
    "\n",
    "df = pd.read_csv('data/input/fourdbs_hi_input_species.csv')\n",
    "# df[~df['species_name'].isin(soi)].reset_index(drop=True).to_csv('data/input/fourdbs_hi_without_3soi.csv', index=False)\n",
    "# df[df['species_name'].isin(soi)].reset_index(drop=True).to_csv('data/input/3soi.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run allegro on without_3soi from above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from Bio import SeqIO\n",
    "\n",
    "from src.utils.guide_finder import GuideFinder\n",
    "\n",
    "\n",
    "def find_all_matches(list1, list2):\n",
    "    matches = list()\n",
    "\n",
    "    for index, item in enumerate(list2):\n",
    "        if item in list1:\n",
    "            matches.append(index)\n",
    "        \n",
    "    return matches\n",
    "\n",
    "def hamming_distance(str1: str, str2: str, length: int) -> int:\n",
    "    \"\"\"\n",
    "    Calculates the Hamming distance between two strings up to a specified length.\n",
    "    \"\"\"\n",
    "    if len(str1) < length or len(str2) < length:\n",
    "        raise ValueError('Strings must have a length of at least {n}'.format(n=length))\n",
    "\n",
    "    return sum(ch1 != ch2 for ch1, ch2 in zip(str1[:length], str2[:length]))\n",
    "\n",
    "\n",
    "def get_record_metadata(record):\n",
    "    # This gene's own name -- Usually N/A for unannotated CDS files or chromosomes/scaffolds\n",
    "    gene_match = re.search(gene_regex, record.description)\n",
    "    gene_name = gene_match.group(1) if gene_match is not None else 'N/A'\n",
    "\n",
    "    # This gene's own protein id e.g., XP_022674739.1\n",
    "    protein_id_match = re.search(protein_id_regex, record.description)\n",
    "    protein_id = protein_id_match.group(1) if protein_id_match is not None else 'N/A'\n",
    "\n",
    "    # For example, [orthologous_to_ref_protein=XP_022674739.1], extracts XP_022674739.1\n",
    "    ortho_prot_to_match = re.search(orthologous_protein_regex, record.description)\n",
    "    ortho_prot_id = ortho_prot_to_match.group(1) if ortho_prot_to_match is not None else 'N/A'\n",
    "\n",
    "    # For example, [orthologous_to_gene=HIS7], extracts HIS7\n",
    "    ortho_gene_to_match = re.search(orthologous_name_regex, record.description)\n",
    "    ortho_gene_name = ortho_gene_to_match.group(1) if ortho_gene_to_match is not None else 'N/A'\n",
    "\n",
    "    return gene_name, protein_id, ortho_prot_id, ortho_gene_name\n",
    "\n",
    "\n",
    "gene_regex = r'\\[gene=(.*?)\\]'\n",
    "tag_regex = r'\\[locus_tag=(.*?)\\]'\n",
    "protein_id_regex = r'\\[protein_id=(.*?)\\]'\n",
    "reference_species_regex = r'\\[ref_species=(.*?)\\]'\n",
    "orthologous_name_regex = r'\\[orthologous_to_gene=(.*?)\\]'\n",
    "orthologous_protein_regex = r'\\[orthologous_to_ref_protein=(.*?)\\]'\n",
    "\n",
    "new_exp_name = 'all_without_threeSOI_a7'\n",
    "\n",
    "split_out = pd.read_csv(f'data/output/{new_exp_name}/{new_exp_name}.csv')\n",
    "\n",
    "library = split_out.sequence.unique()\n",
    "\n",
    "test_df = pd.read_csv(f'data/input/3soi.csv')[['species_name', 'cds_file_name']]\n",
    "\n",
    "df_species_name_list = list()\n",
    "df_covered_list = list()\n",
    "df_strands = list()\n",
    "df_mismatch = list()\n",
    "df_locations = list()\n",
    "df_gene_names = list()\n",
    "df_protein_ids = list()\n",
    "df_ortho_prot_ids = list()\n",
    "df_ortho_gene_names = list()\n",
    "df_lib_partial_match = list()\n",
    "\n",
    "gf = GuideFinder()\n",
    "base_path = 'data/input/cds/orthogroups/'\n",
    "\n",
    "for label, row in test_df.iterrows():\n",
    "    species_name = row['species_name']\n",
    "\n",
    "    species_is_covered_n_times = 0\n",
    "    exact_matching_guides = list()\n",
    "\n",
    "    records_path = base_path + row['cds_file_name']\n",
    "    records = list(SeqIO.parse(open(records_path), 'fasta'))\n",
    "\n",
    "    for record in records:\n",
    "        gene_name, protein_id, ortho_prot_id, ortho_gene_name = get_record_metadata(record)\n",
    "\n",
    "        (guides_list,\n",
    "        _guides_context_list,\n",
    "        strands_list,\n",
    "        locations_list) = gf.identify_guides_and_indicate_strand(\n",
    "            pam='NGG',\n",
    "            sequence=str(record.seq).upper(),\n",
    "            protospacer_length=20,\n",
    "            context_toward_five_prime=0,\n",
    "            context_toward_three_prime=0,\n",
    "            filter_repetitive=False\n",
    "        )\n",
    "\n",
    "        # Attempt to find an exact match.\n",
    "        matches = find_all_matches(library, guides_list)\n",
    "        for match in matches:\n",
    "            df_species_name_list.append(species_name)\n",
    "            df_covered_list.append(guides_list[match])\n",
    "            df_locations.append(locations_list[match])\n",
    "            df_strands.append(strands_list[match])\n",
    "            df_gene_names.append(gene_name)\n",
    "            df_protein_ids.append(protein_id)\n",
    "            df_ortho_prot_ids.append(ortho_prot_id)\n",
    "            df_ortho_gene_names.append(ortho_gene_name)\n",
    "            df_lib_partial_match.append('N/A')\n",
    "            df_mismatch.append('N/A')\n",
    "            species_is_covered_n_times += 1\n",
    "            exact_matching_guides.append(match)\n",
    "    \n",
    "    # Attempt to find partial matches.\n",
    "    mm_allowed = 2\n",
    "    req_match_len = 10\n",
    "\n",
    "    for record in records:\n",
    "        gene_name, protein_id, ortho_prot_id, ortho_gene_name = get_record_metadata(record)\n",
    "\n",
    "        (guides_list,\n",
    "        _guides_context_list,\n",
    "        strands_list,\n",
    "        locations_list) = gf.identify_guides_and_indicate_strand(\n",
    "            pam='NGG',\n",
    "            sequence=str(record.seq).upper(),\n",
    "            protospacer_length=20,\n",
    "            context_toward_five_prime=0,\n",
    "            context_toward_three_prime=0,\n",
    "            filter_repetitive=False\n",
    "        )\n",
    "\n",
    "        for idx, test_guide in enumerate(guides_list):\n",
    "            for lib_guide in library:\n",
    "\n",
    "                # do not allow double-counting exact matching guides from before.\n",
    "                if lib_guide in exact_matching_guides:\n",
    "                    continue\n",
    "\n",
    "                distance_after_seed = hamming_distance(lib_guide, test_guide, req_match_len)\n",
    "                \n",
    "                if lib_guide[req_match_len:] == test_guide[req_match_len:] and distance_after_seed <= mm_allowed and distance_after_seed > 0:\n",
    "                    df_species_name_list.append(species_name)\n",
    "                    df_covered_list.append(test_guide)\n",
    "                    df_locations.append(locations_list[idx])\n",
    "                    df_strands.append(strands_list[idx])\n",
    "                    df_gene_names.append(gene_name)\n",
    "                    df_protein_ids.append(protein_id)\n",
    "                    df_ortho_prot_ids.append(ortho_prot_id)\n",
    "                    df_ortho_gene_names.append(ortho_gene_name)\n",
    "                    df_mismatch.append(distance_after_seed)\n",
    "                    df_lib_partial_match.append(lib_guide)\n",
    "                    species_is_covered_n_times += 1\n",
    "\n",
    "    # No match.\n",
    "    if species_is_covered_n_times == 0:\n",
    "        df_species_name_list.append(species_name)\n",
    "        df_covered_list.append('Not covered')\n",
    "        df_strands.append('Not covered')\n",
    "        df_locations.append('Not covered')\n",
    "        df_gene_names.append('Not covered')\n",
    "        df_protein_ids.append('Not covered')\n",
    "        df_ortho_prot_ids.append('Not covered')\n",
    "        df_ortho_gene_names.append('Not covered')\n",
    "        df_mismatch.append('Not covered')\n",
    "        df_lib_partial_match.append('Not covered')\n",
    "\n",
    "test_df = pd.DataFrame.from_dict({\n",
    "    'species': df_species_name_list,\n",
    "    'covered_by': df_covered_list,\n",
    "    'partial_match_in_library': df_lib_partial_match,\n",
    "    'mismatch': df_mismatch,\n",
    "    'strand': df_strands,\n",
    "    'location': df_locations,\n",
    "    'gene_name': df_gene_names,\n",
    "    'protein_id': df_protein_ids,\n",
    "    'ortho_prot_id': df_ortho_prot_ids,\n",
    "    'ortho_gene_name': df_ortho_gene_names,\n",
    "})\n",
    "\n",
    "test_df.to_csv(f'data/output/test_{new_exp_name}_results.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>covered_by</th>\n",
       "      <th>partial_match_in_library</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>strand</th>\n",
       "      <th>location</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ortho_prot_id</th>\n",
       "      <th>ortho_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CATCTCCAGATGATTGCCAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500977.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CTCCAGATGATTGCCATCGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500977.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>ATGACGACGTTCATGACAGA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RC</td>\n",
       "      <td>687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_500977.1</td>\n",
       "      <td>NP_012965.3</td>\n",
       "      <td>GAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>CACTCTGCCACCAAGTGGAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>626</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_503263.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>yarrowia_lipolytica</td>\n",
       "      <td>TCTGCCACCAAGTGGATCGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F</td>\n",
       "      <td>629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XP_503263.1</td>\n",
       "      <td>NP_013406.1</td>\n",
       "      <td>MET17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>TTGTACATACCTTTACCAGC</td>\n",
       "      <td>TTCTTCATACCTTTACCAGC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>RC</td>\n",
       "      <td>2394</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GTGAACATATCTCTTTGGAT</td>\n",
       "      <td>GTAAACATATCTCTTTGGAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RC</td>\n",
       "      <td>2769</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GTGAACATATCTCTTTGGAT</td>\n",
       "      <td>GTGAACATGTCTCTTTGGAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RC</td>\n",
       "      <td>2769</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GTGAACATATCTCTTTGGAT</td>\n",
       "      <td>GTGAAGATATCTCTTTGGAT</td>\n",
       "      <td>1.0</td>\n",
       "      <td>RC</td>\n",
       "      <td>2769</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>kluyveromyces_marxianus</td>\n",
       "      <td>GTGAACATATCTCTTTGGAT</td>\n",
       "      <td>GTGAAGATGTCTCTTTGGAT</td>\n",
       "      <td>2.0</td>\n",
       "      <td>RC</td>\n",
       "      <td>2769</td>\n",
       "      <td>LYS2</td>\n",
       "      <td>XP_022678348.1</td>\n",
       "      <td>NP_009673.1</td>\n",
       "      <td>LYS2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>153 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     species            covered_by partial_match_in_library  \\\n",
       "0        yarrowia_lipolytica  CATCTCCAGATGATTGCCAT                      NaN   \n",
       "1        yarrowia_lipolytica  CTCCAGATGATTGCCATCGG                      NaN   \n",
       "2        yarrowia_lipolytica  ATGACGACGTTCATGACAGA                      NaN   \n",
       "3        yarrowia_lipolytica  CACTCTGCCACCAAGTGGAT                      NaN   \n",
       "4        yarrowia_lipolytica  TCTGCCACCAAGTGGATCGG                      NaN   \n",
       "..                       ...                   ...                      ...   \n",
       "148  kluyveromyces_marxianus  TTGTACATACCTTTACCAGC     TTCTTCATACCTTTACCAGC   \n",
       "149  kluyveromyces_marxianus  GTGAACATATCTCTTTGGAT     GTAAACATATCTCTTTGGAT   \n",
       "150  kluyveromyces_marxianus  GTGAACATATCTCTTTGGAT     GTGAACATGTCTCTTTGGAT   \n",
       "151  kluyveromyces_marxianus  GTGAACATATCTCTTTGGAT     GTGAAGATATCTCTTTGGAT   \n",
       "152  kluyveromyces_marxianus  GTGAACATATCTCTTTGGAT     GTGAAGATGTCTCTTTGGAT   \n",
       "\n",
       "     mismatch strand  location gene_name      protein_id ortho_prot_id  \\\n",
       "0         NaN      F       224       NaN     XP_500977.1   NP_012965.3   \n",
       "1         NaN      F       227       NaN     XP_500977.1   NP_012965.3   \n",
       "2         NaN     RC       687       NaN     XP_500977.1   NP_012965.3   \n",
       "3         NaN      F       626       NaN     XP_503263.1   NP_013406.1   \n",
       "4         NaN      F       629       NaN     XP_503263.1   NP_013406.1   \n",
       "..        ...    ...       ...       ...             ...           ...   \n",
       "148       2.0     RC      2394      LYS2  XP_022678348.1   NP_009673.1   \n",
       "149       1.0     RC      2769      LYS2  XP_022678348.1   NP_009673.1   \n",
       "150       1.0     RC      2769      LYS2  XP_022678348.1   NP_009673.1   \n",
       "151       1.0     RC      2769      LYS2  XP_022678348.1   NP_009673.1   \n",
       "152       2.0     RC      2769      LYS2  XP_022678348.1   NP_009673.1   \n",
       "\n",
       "    ortho_gene_name  \n",
       "0              GAP1  \n",
       "1              GAP1  \n",
       "2              GAP1  \n",
       "3             MET17  \n",
       "4             MET17  \n",
       "..              ...  \n",
       "148            LYS2  \n",
       "149            LYS2  \n",
       "150            LYS2  \n",
       "151            LYS2  \n",
       "152            LYS2  \n",
       "\n",
       "[153 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(f'data/output/test_{new_exp_name}_results.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allegro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
